misc/copyright.py

```diff
@@ -0,0 +1,302 @@
+    if not dry_run:
+        with open(filepath, 'w') as f:
+            if new_header_lines:
+                f.writelines(new_header_lines)
+            f.writelines(body_lines)
+    return return_state
+
+
+class WorkStats:
+    def __init__(self):
+        self.opened_file_num = 0
+        self.inserted_notice_file_num = 0
+        self.modified_notice_file_num = 0
+
+
+# Available on POSIX, Windows.
+try:
+    LINE_ELIDING = os.get_terminal_size().columns >= 72
+except AttributeError:  # Maybe we want to run on niche platforms..
+    LINE_ELIDING = False
+
+
+PLAYFUL_BRAILLE = ["⠃", "⠆", "⠤", "⠰", "⠘", "⠉"]
+
+
+def print_progress(stats: WorkStats, dry_run: bool):
+    content = "{dots} Opened {opened}, {tense} insert notice: {insert}, {tense} modify notice: {modify}".format(
+        opened=stats.opened_file_num,
+        tense="will" if dry_run else "did",
+        insert=stats.inserted_notice_file_num,
+        modify=stats.modified_notice_file_num,
+        dots=PLAYFUL_BRAILLE[stats.opened_file_num % len(PLAYFUL_BRAILLE)])
+    # 1A, 2K: move cursor one line up and clear the entire line.
+    sys.stdout.write(("\x1b[1A\x1b[2K" if LINE_ELIDING else "") + content + "\n")
+
+
+def work_on_file(filepath: str, ext: str, stats: WorkStats, dry_run: bool):
+    stats.opened_file_num += 1
+    status = check_and_modify(filepath, FILE_EXT_TO_COMMENT_STYLES[ext], dry_run)
+    if status == FileActionResult.INSERTED_NOTICE:
+        stats.inserted_notice_file_num += 1
+    elif status == FileActionResult.MODIFIED_NOTICE:
+        stats.modified_notice_file_num += 1
+    print_progress(stats, dry_run)
+    return
+
+
+def is_interested_ext(ext: str, selected_stripped_exts: List[str]) -> bool:
+    """
+    Returns whether we are interested in a file based on its extension name.
+    NOTE ext contains a dot, e.g. ".cpp", ".sh", while selected_stripped_exts elements
+         do not, e.g. [ "cpp", "sh" ]
+    """
+    return (ext in FILE_EXT_TO_COMMENT_STYLES) \
+        and ((not selected_stripped_exts) or (ext.lstrip(".") in selected_stripped_exts))
+
+
+def work(args):
+    start_time = time.time()
+    stats = WorkStats()
+    picked_exts = [] if not args.exts else args.exts.split(",")
+    print("Starting.")
+    for path in args.paths:
+        if os.path.isdir(path):
+            for (dirpath, dirnames, filenames) in os.walk(path):
+                for f in filenames:
+                    ext = os.path.splitext(f)[-1]
+                    if not is_interested_ext(ext, picked_exts):
+                        continue
+                    work_on_file(os.path.join(dirpath, f), ext, stats,
+                                 args.dry_run)
+        elif os.path.isfile(path):
+            ext = os.path.splitext(path)[-1]
+            if not is_interested_ext(ext, picked_exts):
+                continue
+            work_on_file(path, ext, stats, args.dry_run)
+    print("Done in %.1f sec." % (time.time() - start_time))
```

Comment:
Can we add a return value here, so that we can know if all files satisfy the code format? With this feature we can add this script to our GitHub action so that the copyright info will be enforced in all files. This also makes L295 less confusing.

Reply:
Sure, will incorporate in a new commit.

Reply:
I think we can repurpose the `--dry-run` option here.